Mathlib/Combinatorics/Optimization/ValuedCSP.lean

```diff
@@ -0,0 +1,79 @@
+
+General-Valued CSP is a very broad class of problems in discrete optimization.
+General-Valued CSP subsumes Min-Cost-Hom (including 3-SAT for example) and Finite-Valued CSP.
+
+## Main definitions
+* `ValuedCspTemplate`: A VCSP template; fixes a domain, a codomain, and allowed cost functions.
+* `ValuedCspTerm`: One summand in a VCSP instance; calls a concrete function from given template.
+* `ValuedCspTerm.evalSolution`: An evaluation of the VCSP term for given solution.
+* `ValuedCspInstance`: An instance of a VCSP problem over given template.
+* `ValuedCspInstance.evalSolution`: An evaluation of the VCSP instance for given solution.
+* `ValuedCspInstance.optimumSolution`: Is given solution a minimum of the VCSP instance?
+
+## References
+* [D. A. Cohen, M. C. Cooper, P. Creed, P. G. Jeavons, S. Živný,
+   *An Algebraic Theory of Complexity for Discrete Optimisation*][cohen2012]
+
+-/
+
+def n1aryOfUnary {α β : Type*} (f : α → β) : (Fin 1 → α) → β :=
+  fun a => f (a 0)
+
+def n2aryOfBinary {α β : Type*} (f : α → α → β) : (Fin 2 → α) → β :=
+  fun a => f (a 0) (a 1)
+
+/-- A template for a valued CSP problem over a domain `D` with costs in `C`. -/
+@[reducible]
+def ValuedCspTemplate (D C : Type*) [LinearOrderedAddCommMonoid C] :=
+  Set (Σ (n : ℕ), (Fin n → D) → C) -- Cost functions `D^n → C` for any `n`
+
+variable {D C : Type*} [LinearOrderedAddCommMonoid C]
+
+/-- A term in a valued CSP instance over the template `Γ`. -/
+structure ValuedCspTerm (Γ : ValuedCspTemplate D C) (ι : Type*) where
+  /-- Arity of the function -/
+  n : ℕ
+  /-- Which cost function is instantiated -/
+  f : (Fin n → D) → C
+  /-- The cost function comes from the template -/
+  inΓ : ⟨n, f⟩ ∈ Γ
+  /-- Which variables are plugged as arguments to the cost function -/
+  app : Fin n → ι
+
+def valuedCspTermOfUnary {Γ : ValuedCspTemplate D C} {ι : Type*} {f₁ : D → C}
+    (ok : ⟨1, n1aryOfUnary f₁⟩ ∈ Γ) (i : ι) : ValuedCspTerm Γ ι :=
+  ⟨1, n1aryOfUnary f₁, ok, ![i]⟩
+
+def valuedCspTermOfBinary {Γ : ValuedCspTemplate D C} {ι : Type*} {f₂ : D → D → C}
+    (ok : ⟨2, n2aryOfBinary f₂⟩ ∈ Γ) (i j : ι) : ValuedCspTerm Γ ι :=
+  ⟨2, n2aryOfBinary f₂, ok, ![i, j]⟩
+
+/-- Evaluation of a `Γ` term `t` for given solution `x`. -/
+def ValuedCspTerm.evalSolution {Γ : ValuedCspTemplate D C} {ι : Type*}
+    (t : ValuedCspTerm Γ ι) (x : ι → D) : C :=
+  t.f (x ∘ t.app)
+
+/-- A valued CSP instance over the template `Γ` with variables indexed by `ι`.-/
+def ValuedCspInstance (Γ : ValuedCspTemplate D C) (ι : Type*) :=
+  List (ValuedCspTerm Γ ι)
+
+/-- Evaluation of a `Γ` instance `I` for given solution `x`. -/
+def ValuedCspInstance.evalSolution {Γ : ValuedCspTemplate D C} {ι : Type*}
+    (I : ValuedCspInstance Γ ι) (x : ι → D) : C :=
+  (I.map (·.evalSolution x)).sum
+
+/-- Condition for `x` being an optimum solution (min) to given `Γ` instance `I`.-/
+def ValuedCspInstance.OptimumSolution {Γ : ValuedCspTemplate D C} {ι : Type*}
```

Comment:
```suggestion
def ValuedCspInstance.OptimalSolution {Γ : ValuedCspTemplate D C} {ι : Type*}
```
`Optimum` is usually used as a noun.

Reply:
I thought that `Optimum` is stronger than `Optimal`, i.e., `Optimum` is better than everything, `Optimal` can be incomparable with some elements (you can have several nonidentical optimal solutions). Did I get the terminology wrong?

Reply:
I'm not sure, it's not important though.